Feature/dev harrel impl

harrel56 opened the pull request:
### Reviewer checklist
 - [ ] Read the [contributing guide](https://github.com/creek-service/.github/blob/main/CONTRIBUTING.md)
 - [ ] PR should be motivated, i.e. what does it fix, why, and if relevant, how
 - [ ] Ensure relevant issues are linked (description should include text like "Fixes #<issue number>")
 - [ ] Ensure any appropriate documentation has been added or amended

harrel56 commented:
Hi, let me know if there is something missing. I have no suggestion though: could you somehow separate spec compliance to required & optional? Because now it feels a little bit unfair as I have the highest score for required and the lowest score for overall. I just don't support format assertions (yet?) which is most of the optional tests are checking.

Also I think the serde tests might be a bit unfair? E.g. "Medeia" doesn't support newest drafts so it is run with draft7 and it cannot really be compared to implementations which are run with draft2020-12. Probably this also should be grouped by specs?

harrel56 commented:
And just out of curiosity: how long does it take you to run all benchmarks? :P It looks like it might take a whole night or so

big-andy-coates commented:
Thanks @harrel56, I'll take a look at this.

> could you somehow separate spec compliance to required & optional? 

Results are split between required and optional. Though currently the graph is not. 

The table summarising the functional results breaks them down into pass/fail counts on **o**ptional and **r**required features.

The over all **score** treats blends both required and optional, with required test cases making up 75% of the score.

It would be possible to add a graph comparing only **required** functional coverage. Feel free to raise an issue, and ideally work on it.

> Also I think the serde tests might be a bit unfair? E.g. "Medeia" doesn't support newest drafts so it is run with draft7 and it cannot really be compared to implementations which are run with draft2020-12. 

Yes, I'm aware of this.

I'm currently working on making the charts and tables of results be driven from the latest test runs, so that results update as new versions of validator libraries come out.  As part of this work, I have a `todo` to call out which draft version the serde benchmark is using, what this means, and to potentially break it down by spec.

Time is always a limiting factor though ;)

harrel56 commented:
When the benchmarks and graphs are going to be automatically generated I'll be happy to help out with splitting this graph - I just don't wanna do it manually. Gonna report this as a separate issue then

big-andy-coates commented on src/main/java/org/creekservice/kafka/test/perf/implementations/DevHarrelImplementation.java:
```diff
@@ -0,0 +1,146 @@
+    private static final MetaData METADATA =
+            new MetaData(
+                    "json-schema (dev.harrel)",
+                    "json-schema (dev.harrel)",
```
Shorter name works better in tables & charts.

```suggestion
                    "dev.harrel",
```

big-andy-coates commented on src/main/java/org/creekservice/kafka/test/perf/implementations/DevHarrelImplementation.java:
```diff
@@ -0,0 +1,146 @@
+                    "json-schema (dev.harrel)",
+                    Language.Java,
+                    Licence.MIT,
+                    Set.of(DRAFT_2020_12, DRAFT_2019_09),
```
I'm sure you're aware, but currently out of sync with https://json-schema.org/implementations#java, which states only 2020_12.

harrel56 replied:
Yeah, support for 2019-09 is a new addition - I have a pending PR for their site which should update it

big-andy-coates commented on src/main/java/org/creekservice/kafka/test/perf/implementations/DevHarrelImplementation.java:
```diff
@@ -0,0 +1,146 @@
+    private final URI testSchemaUri = URI.create("urn:test");
+    private Map<URI, String> remotes = Map.of();
+
+    public DevHarrelImplementation() {
+        final SchemaResolver schemaResolver =
+                uri -> {
+                    final String resolved = remotes.get(URI.create(uri));
+                    if (resolved == null) {
+                        return SchemaResolver.Result.empty();
+                    }
+                    return SchemaResolver.Result.fromString(resolved);
+                };
+        final Validator validator2020 =
+                new dev.harrel.jsonschema.ValidatorFactory()
+                        .withSchemaResolver(schemaResolver)
+                        .createValidator();
+        final Validator validator2019 =
+                new dev.harrel.jsonschema.ValidatorFactory()
+                        .withDialect(new Dialects.Draft2019Dialect())
+                        .withSchemaResolver(schemaResolver)
+                        .createValidator();
+        /* Validate against meta-schemas in order to parse them eagerly */
+        validator2020.validate(URI.create(SpecificationVersion.DRAFT2020_12.getId()), "{}");
+        validator2019.validate(URI.create(SpecificationVersion.DRAFT2019_09.getId()), "{}");
+
+        validator2020.registerSchema(testSchemaUri, TestSchemas.DRAFT_2020_SCHEMA);
+
+        this.validators =
+                Map.of(
+                        DRAFT_2020_12, validator2020,
+                        DRAFT_2019_09, validator2019);
+    }
```
My fault for moving the goal posts on you, but...

You shouldn't need anything in the constructor any more.  Instead, the work is done in `prepare`.  

Take a look at other impls. Look no constructors!  

`prepare` is _not_ included in performance benchmarks.

big-andy-coates commented on src/main/java/org/creekservice/kafka/test/perf/implementations/DevHarrelImplementation.java:
```diff
@@ -0,0 +1,146 @@
+    @Override
+    public JsonValidator prepare(
+            final String schema, final SchemaSpec spec, final AdditionalSchemas additionalSchemas) {
+        DevHarrelImplementation.this.remotes = additionalSchemas.remotes();
+        final Validator validator = validators.get(spec);
+        final URI schemaUri = validator.registerSchema(schema);
```
With the constructor gone, you'll need something like:

```suggestion
        final Validator validator = validator(spec, additionalSchemas);       
        final URI schemaUri = validator.registerSchema(schema);
```

With:

```java
private Validator validator(final SchemaSpec spec, final AdditionalSchemas additionalSchemas) {
   final SchemaResolver schemaResolver =
                uri -> {
                    final String resolved = additionalSchemas.remotes().get(URI.create(uri));
                    if (resolved == null) {
                        return SchemaResolver.Result.empty();
                    }
                    return SchemaResolver.Result.fromString(resolved);
                };

    switch (spec) {
      case: SchemaSpec.DRAFT_2020_12:
           final Validator validator = new dev.harrel.jsonschema.ValidatorFactory()
                        .withSchemaResolver(schemaResolver)
                        .createValidator();
            /* Validate against meta-schemas in order to parse them eagerly */
           // Note: If `registerSchema()` will cause this to be parsed, then I don't think this will be necessary, as schema parsing is not currently part of performance benchmark
            validator.validate(URI.create(SpecificationVersion.DRAFT2020_12.getId()), "{}");
            return validator;
      case: SchemaSpec.DRAFT_2019_09:
           final Validator validator = new dev.harrel.jsonschema.ValidatorFactory()
                        .withDialect(new Dialects.Draft2019Dialect())
                        .withSchemaResolver(schemaResolver)
                        .createValidator();
            /* Validate against meta-schemas in order to parse them eagerly */
             // Note: If `registerSchema()` will cause this to be parsed, then I don't think this will be necessary, as schema parsing is not currently part of performance benchmark

            validator.validate(URI.create(SpecificationVersion.DRAFT2019_09.getId()), "{}");
            return validator;
   default:
      throw new RuntimeException("Unsupported Spec:" + spec);
}
```

harrel56 replied:
Unfortunately meta-schemas will only get parsed when needed for the first time - with the first validation

big-andy-coates commented on src/main/java/org/creekservice/kafka/test/perf/implementations/DevHarrelImplementation.java:
```diff
@@ -0,0 +1,146 @@
+            @Override
+            public TestModel deserialize(final byte[] data) {
+                try {
+                    if (!validators
+                            .get(DRAFT_2020_12)
+                            .validate(testSchemaUri, new String(data, StandardCharsets.UTF_8))
+                            .isValid()) {
+                        throw new RuntimeException();
+                    }
+                    return mapper.readValue(data, TestModel.class);
```
Might be quicker to avoid double string decoding:


```suggestion
                    final String json = new String(data, StandardCharsets.UTF_8);
                    if (!validators
                            .get(DRAFT_2020_12)
                            .validate(testSchemaUri, json)
                            .isValid()) {
                        throw new RuntimeException();
                    }
                    return mapper.readValue(json, TestModel.class);
```

harrel56 replied:
yep, good point

big-andy-coates commented on src/main/java/org/creekservice/kafka/test/perf/implementations/DevHarrelImplementation.java:
```diff
@@ -0,0 +1,146 @@
+                                    .get(DRAFT_2020_12)
+                                    .validate(testSchemaUri, asString)
+                                    .isValid()) {
+                        throw new RuntimeException();
```
Does your library expose details of _why_ it failed?  

Can be useful to expose these in the exception, as it is logged...

harrel56 replied:
If I understand correctly I am required to throw if the validation fails, right? Is it being logged even if it's expected to fail?

big-andy-coates replied:
> If I understand correctly I am required to throw if the validation fails, right?

Correct.

> Is it being logged even if it's expected to fail?

I've not checked the code, but I wouldn't of thought so. It should only be logged if the test was supposed to pass.  Exposing the error message can make it easier to track down the cause of false negatives.

harrel56 replied:
Hmm, I just don't want to add more overhead to the validation logic. Gonna try to make as lightweight as it can be

big-andy-coates commented on src/main/java/org/creekservice/kafka/test/perf/implementations/DevHarrelImplementation.java:
```diff
@@ -0,0 +1,146 @@
+                try {
+                    final String asString = mapper.writeValueAsString(model);
+                    if (validate
+                            && !validators
+                                    .get(DRAFT_2020_12)
+                                    .validate(testSchemaUri, asString)
+                                    .isValid()) {
```
Should now be just:

```suggestion
                            && !validator
                                    .validate(testSchemaUri, asString)
                                    .isValid()
```

big-andy-coates commented on src/main/java/org/creekservice/kafka/test/perf/implementations/DevHarrelImplementation.java:
```diff
@@ -0,0 +1,146 @@
+            @Override
+            public TestModel deserialize(final byte[] data) {
+                try {
+                    if (!validators
+                            .get(DRAFT_2020_12)
+                            .validate(testSchemaUri, new String(data, StandardCharsets.UTF_8))
+                            .isValid()) {
```
Should now be just"

```suggestion
                    if (!validator
                            .validate(testSchemaUri, new String(data, StandardCharsets.UTF_8))
                            .isValid()) {
```

big-andy-coates commented on src/main/java/org/creekservice/kafka/test/perf/testsuite/TestSuite.java:
```diff
@@ -53,7 +54,10 @@ public TestSuite(
             this.optional =
                     suiteFilePath.getParent() != null
-                            && suiteFilePath.getParent().toString().contains("/optional");
+                            && suiteFilePath
+                                    .getParent()
+                                    .toString()
+                                    .contains(File.separator + "optional");
```
Nice. Thanks.

big-andy-coates commented on src/main/java/org/creekservice/kafka/test/perf/testsuite/TestSuiteLoader.java:
```diff
@@ -94,7 +94,10 @@ public JsonSchemaTestSuite load(final Path rootDir) {
 
         final Function<Path, URI> createKey =
-                path -> URI.create("http://localhost:1234/" + remotes.relativize(path));
+                path ->
+                        URI.create(
+                                "http://localhost:1234/"
+                                        + remotes.relativize(path).toString().replace("\\", "/"));
```
Thanks!

big-andy-coates commented on src/test/java/org/creekservice/kafka/test/perf/implementations/DevHarrelImplementationTest.java:
```diff
@@ -0,0 +1,19 @@
+
+package org.creekservice.kafka.test.perf.implementations;
+
+class DevHarrelImplementationTest extends ImplementationTest {}
```
Class no longer needed (again, this is me moving the goal posts on you!)

harrel56 replied:
Hmm, how so? I can still see test classes for other implementations

big-andy-coates replied:
Sorry, I'm tripping - thought this was a different class.

big-andy-coates reviewed (commented):
Hey @harrel56,

First off, a big apology for committing a refactor while you were in the middle of your work.  My bad.  I'm just trying to finish off a few things to make results more data driven.

This looks great! 

I've made a few comments below.  Mainly looks like stuff due to you refactoring the code after I changed things.  A bit of clean up and should be good to go.

big-andy-coates commented on src/main/java/org/creekservice/kafka/test/perf/implementations/DevHarrelImplementation.java:
```diff
@@ -0,0 +1,146 @@
+                    Language.Java,
+                    Licence.MIT,
+                    Set.of(DRAFT_2020_12, DRAFT_2019_09),
+                    "https://github.com/harrel56/json-schema");
```
Hey @harrel56,

I'm about to merge another PR: https://github.com/creek-service/json-schema-validation-comparison/pull/54

This should NOT stop your code compiling (I hope).  However, it would be great if you could update your PR as I've noted below to set a color for your impl in the result charts.

,

```suggestion
                    "https://github.com/harrel56/json-schema",
                     new Color(235, 54, 172));
```

harrel56 replied:
Changed the color to something else - felt like the pinky one was already taken? Honestly hard to tell how it's going to look like on the charts. BTW this chart page look really nice

big-andy-coates replied:
You can build them locally, but I've not yet documented how to.  Feel free to change the color in a new PR if you don't like it.

big-andy-coates commented:
Sorry - forgot I moved those files!

harrel56 commented:
> First off, a big apology for committing a refactor while you were in the middle of your work. My bad. I'm just trying to finish off a few things to make results more data driven.

Yeah, that really wasn't a problem so no worries :)

I think I addressed all your comments - let me know if there's anything left

big-andy-coates commented on src/main/java/org/creekservice/kafka/test/perf/implementations/DevHarrelImplementation.java:
```diff
@@ -0,0 +1,141 @@
+
+        final Validator validator = validator(spec);
+        /* Parse remotes eagerly and ignore errors from unknown specs */
+        additionalSchemas
+                .remotes()
+                .forEach(
+                        (uri, remote) -> {
+                            try {
+                                validator.registerSchema(uri, remote);
+                            } catch (Exception e) {
+                                /* ignore */
+                            }
+                        });
+        final URI schemaUri = validator.registerSchema(schema);
+
```
Rather than parsing _all_ remotes each time, does the validator not have a way of loading them on demand?  

While this won't _currently_ affect benchmark results, as the prepare statement isn't timed, users may look at this implementation as a reference.  Good practice here is to be encouraged :)

In the future, there could be a benchmark measuring the time it takes to prepare, i.e. parse, the schemas.

harrel56 replied:
Not parsing remotes eagerly will affect benchmark - it would get parsed during validation logic (it's not parsed during registration process, references are resolved only if needed during validation). But I can change it - I checked it and it looked like it wasn't a significant performance hit

big-andy-coates replied:
If that's the case, then I have no issue with you parsing the in prepare.  

Unless there was some way you could force your impl to parse them?  I'm guess trying to validate an empty document in `prepare` wouldn't cause the reference to be loaded?

harrel56 replied:
Currently I cannot think of a way for achieving that, empty document wouldn't trigger that (in most cases). But I tested out the performance for both solutions and honestly I didn't notice any difference - so it's really fine

big-andy-coates reviewed (commented):
LGTM @harrel56 

Few minors below, plus the following remaining from the previous reviews:

- https://github.com/creek-service/json-schema-validation-comparison/pull/50#discussion_r1391218661
- https://github.com/creek-service/json-schema-validation-comparison/pull/50#discussion_r1391202114

harrel56 commented:
Ok, completed another iteration - it's getting a little messy, hopefully didn't omit anything

big-andy-coates reviewed (approved):
LGTM!

Thanks for your contribution.

big-andy-coates commented:
Performance results are up: https://github.com/json-schema-org/website/pull/225
